### DISCUSSION TITLE
File Upload Questions

### DISCUSSION THREAD

**Discussion by ZehuaZhang** [Milkdown/milkdown#988] [Q&A] [answered]
First of all thank you for such amazing markdown editor and preview, and plugin design mindset! 

I looked at your file upload plugin documentation, and tried the playground. Could you please tell me more about the features the file upload plugin has? I assume it supports drag and drop, and also copy and paste to upload to dedicated cloud storage? And does it have a toolbar button on the top where user could click to upload the file from local drives?

**Comment by zgrybus** [comment, marked as answer]
@ZehuaZhang 
> And does it have a toolbar button on the top where user could click to upload the file from local drives

I believe you need to implementing it by yourself. We did it and it's pretty simple :) You just need to call:
```typescript
editor.action(insertImageCommand.key, { src: source })
```
when saving the image.

Here is simple demo, where you may upload a file via menu bar - https://daredata.github.io/notion-style-editor/

**Reply by ZehuaZhang** [reply]
Thank you, I assume the current image functionality is copy and paste the raw binary content directly into the markdown? Is there a cloud storage option, where I could drag and drop save into the cloud storage and return an url.
What format does the markdown support? And besides videos does it support audio, or video clips?
And is that possible for you to share the source code for your demo?

**Reply by zgrybus** [reply]
@ZehuaZhang 
> Is there a cloud storage option, where I could drag and drop save into the cloud storage and return an url.

You may override default behaviour and pass your custom function for saving a file.

You may check this https://github.com/DareData/notion-style-editor/blob/main/src/lib/packages/EditorContext/hooks/useUploadPlugin/useUploader.tsx
Documentation of the Milkdown mentions about it as well
